pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java

```diff
@@ -3411,18 +3415,38 @@ private long readLongValue(byte[] bytes, int oid, long minVal, long maxVal, Stri
         val = ByteConverter.int8(bytes, 0);
         break;
       case Oid.FLOAT4:
-        val = (long) ByteConverter.float4(bytes, 0);
+        float f = ByteConverter.float4(bytes, 0);
+        // for float values we know to be within values of long, just cast directly to long
+        if (f <= LONG_MAX_FLOAT && f >= LONG_MIN_FLOAT) {
+          val = (long) f;
+        } else {
+          throw new PSQLException(GT.tr("Bad value for type {0} : {1}", targetType, f),
+              PSQLState.NUMERIC_VALUE_OUT_OF_RANGE);
+        }
         break;
       case Oid.FLOAT8:
-        val = (long) ByteConverter.float8(bytes, 0);
+        double d = ByteConverter.float8(bytes, 0);
+        // for double values within the values of a long, just directly cast to long
+        if (d <= LONG_MAX_FLOAT && d >= LONG_MIN_FLOAT) {
```

Comment:
presumably we need to use the double variants of nextUp and nextDown for the float8.

https://docs.oracle.com/javase/8/docs/api/java/lang/StrictMath.html#nextUp-double-
https://docs.oracle.com/javase/8/docs/api/java/lang/StrictMath.html#nextDown-double-

Reply:
This is easy enough, thanks